### PR TITLE
Implemented first (proper) version or roleplay economy plugin.

### PR DESCRIPTION
**Pull request by GTG3000**
Features:
- Keeps track of character sheets, including:
  - Fully customiseable fields, meant to store misc data
  - Character inventories (including a "key" inventory that characters can not edit on their own)
  - Money
  - A single image per character
  - Owner of the character may transfer and sell items at whim.
  - Integration with roleplay plugin allows for easy creation of characters to be filled out.
- An item shop, enabling characters to purchase/sell items.
  - Support for fully customizeable unique items to be given to characters, without cluttering up the shop.

Commands:

User commands:
- Char
- Shop
- ShopCategories
- ShopItems
- ItemInfo
- DumpItem

Moderator commands:
- ListChars
- ListItems
- Give
- Take
- GiveCustom
- UploadChar
- UploadItem
- DeleteChar
- DeleteItem

Maintainer commands:
- ReloadEconChars
- ReloadEconItems

### CODE REVIEW

File: roleplay_economy.py [hunk 1]
```diff
@@ -0,0 +1,1098 @@
...
+import json
+import re
+from copy import deepcopy
+from time import time
+
+
+class RoleplayEconomy(BasePlugin):
+    name = "roleplay_economy"
+    version = "1.0"
+    author = "GTG3000"
```
Comment (rubellyte) [inline roleplay_economy.py:262]: Don't forget your `description` metafield

File: roleplay_economy.py [hunk 2]
```diff
@@ -0,0 +1,1098 @@
...
+    }
+
+    item_base = {
+        "name": "Default Item",
+        "description": "This item looks very generic",
+        "category": None,
+        "image": None,
+        "inshop": False,
+        "p_buy": 100,
+        "p_sell": 100,
```
Comment (rubellyte) [inline roleplay_economy.py]: I feel like these should probably be a bit more verbose, something like "buy_price" and "sell_price"

File: roleplay_economy.py [hunk 3]
```diff
@@ -0,0 +1,1098 @@
...
+    # technically, if the bot is shut down properly, we don't need to save them manually at all - but you never know.
+    # not doing the same for items because frankly I don't see any reason for items to be modified so often.
+    _save_chars = False
+
+    class Character:
+        _parent = None
+        _chars = None
+        shop = None
+        gid = None
+        id = None
```
Comment (rubellyte) [inline roleplay_economy.py]: I feel like these should be instance attributes rather than class attributes

File: roleplay_economy.py [hunk 4]
```diff
@@ -0,0 +1,1098 @@
...
+
+        def override_item(self, item: dict):
+            return {**self.shop[item['name']], **item['override'],
+                    'fields': [(f, v) for f, v in
+                               list({**dict(self.shop[item['name']]['fields']),
+                                     **dict(item['override'].get('fields', []))}.items())
+                               if v]}
+
+        def embed(self):
+            t_embed = Embed(type="rich", colour=16711680)
```
Comment (rubellyte) [inline roleplay_economy.py]: Perhaps make the color int here a hex literal for easier interpretation?

File: roleplay_economy.py [hunk 5]
```diff
@@ -0,0 +1,1098 @@
...
+            t_embed.title = self.name
+            if self.image:
+                t_embed.set_thumbnail(url=self.image)
+            for (f_name, f_data) in self.fields:
+                t_embed.add_field(name=f_name, value=f_data, inline=False)
+
+            t_str = ""
```
Comment (rubellyte) [inline roleplay_economy.py]: More verbose variable name please

Comment (rubellyte) [reply roleplay_economy.py]: Some comments in this upcoming block to explain what you're doing would also be helpful, it's not particularly readable

File: roleplay_economy.py [hunk 6]
```diff
@@ -0,0 +1,1098 @@
...
+
+        _mpage: int
+        _gid: str
+        _parent = None
+        _message: Message
+
+        _items: list
+
+        _emoji = "◀🇽▶"
+        _len = 10
```
Comment (rubellyte) [inline roleplay_economy.py]: Think these should be instance attributes, same as above

File: roleplay_economy.py [hunk 7]
```diff
@@ -0,0 +1,1098 @@
...
+    def _generate_item_embed(item: dict, custom=False):
+        """
+        Helper function to generate item embeds, given (overriden) item data.
+        The function does not do it's own overriding on account of that requiring it be tied to Character class.
+        :param item:
+        :param custom:
+        :return:
+        """
+
+        item_embed = Embed(type="rich", colour=16711680)
```
Comment (rubellyte) [inline roleplay_economy.py]: blah blah hex literals

File: roleplay_economy.py [hunk 8]
```diff
@@ -0,0 +1,1098 @@
...
+                    output['fields'] = [[str(name)[:32], str(value)[:1024]] for name, value in item['fields']]
+                except (ValueError, TypeError):
+                    raise CommandSyntaxError("Fields must be lists of two values each.")
+            return output
+        except (KeyError, TypeError):
+            raise CommandSyntaxError('Item or Override must be a dict.')
+
+    def _find_item(self, gid: str, query: str):
+        """
+        Function to finda possible item using sequence matcher.
```
Comment (rubellyte) [inline roleplay_economy.py]: finda

File: roleplay_economy.py [hunk 9]
```diff
@@ -0,0 +1,1098 @@
...
+            # char (name) -p (name) (amount)
+            # transfers money between characters, pretty straightforward.
+            try:
+                amount = int(args['pay'][1])
+                if not 0 < amount < char.money:
+                    raise ValueError
+            except ValueError:
+                raise CommandSyntaxError(f"{args['pay'][1]} is not a valid amount.")
+
+            recipient = self.Character(self, gid, args['pay'][0].lower())
```
Comment (rubellyte) [inline roleplay_economy.py]: Why is a new Character object being created here?

File: roleplay_economy.py [hunk 10]
```diff
@@ -0,0 +1,1098 @@
...
+        try:
+            args = parser.parse_args(shlex.split(msg.clean_content))
+        except ValueError as e:
+            self.logger.warning(f"Unable to split {msg.clean_content}. {e}")
+            raise CommandSyntaxError(e)
+
+        if args['amount'] < 1:
+            raise CommandSyntaxError("Please use the Take command to remove items.")
+
+        char = self.Character(self, gid, args['name'])
```
Comment (rubellyte) [inline roleplay_economy.py]: Same concern as above

File: roleplay_economy.py [hunk 11]
```diff
@@ -0,0 +1,1098 @@
...
+        try:
+            args = parser.parse_args(shlex.split(msg.clean_content))
+        except ValueError as e:
+            self.logger.warning(f"Unable to split {msg.clean_content}. {e}")
+            raise CommandSyntaxError(e)
+
+        if args['amount'] < 1:
+            raise CommandSyntaxError("Please use the Give command to add items.")
+
+        char = self.Character(self, gid, args['name'])
```
Comment (rubellyte) [inline roleplay_economy.py]: ^

File: roleplay_economy.py [hunk 12]
```diff
@@ -0,0 +1,1098 @@
...
+        else:
+            # Otherwise, there's a code block attached. If there isn't, welp.
+            data = re.match("(?P<char>.+?)\s+```.*?(?P<json>{.+}).*```", data, re.DOTALL)
+            char = data['char']
+            try:
+                data = json.loads(data['json'])
+            except Exception as e:
+                raise CommandSyntaxError(f"Not a valid JSON block: {e}.\n")
+
+        char = self.Character(self, gid, char.lower())
```
Comment (rubellyte) [inline roleplay_economy.py]: ^

**Review by rubellyte — CHANGES_REQUESTED** [review]

File: roleplay_economy.py (hunk 6 above)
Comment (rubellyte) [reply roleplay_economy.py]: although perhaps not _emoji and _len

File: roleplay_economy.py (hunk 9 above)
Comment (GTG3000) [reply roleplay_economy.py]: Character objects are used as wrappers for interacting with the dictionaries, and do not actually store the data we work with.

I just feel like it improves clarity when you have `recipient.money += x` in stead of `self.chars[gid][args['pay'][0].lower()]['money'] += x`. The upside is that Character object also does verification and initialization for us.

Comment (rubellyte) [reply roleplay_economy.py]: I don't disagree with you on that point, I was just wondering why it was done like so. Seems reasonable.

File: roleplay_economy.py (hunk 6 above)
Comment (GTG3000) [reply roleplay_economy.py]: For all intents and purposes, those are type hints.

**Review by rubellyte — APPROVED** [review]

**Pull request merged**